[Python] test_cuda_numba_interop fails with numba-cuda

### Describe the bug, including details regarding any error messages, version, and platform.

Integrated CUDA support in Numba is [deprecated](https://numba.readthedocs.io/en/stable/cuda/overview.html#cuda-deprecation-status) in favor of the external [Numba-CUDA](https://nvidia.github.io/numba-cuda/) module, maintained by NVidia.

However, when running our [Numba CUDA interop tests](https://github.com/apache/arrow/blob/main/python/pyarrow/tests/test_cuda_numba_interop.py) with Numba-CUDA installed, I get the following failures on all tests:
```
  File "/home/antoine/arrow/dev/python/pyarrow/tests/test_cuda_numba_interop.py", line 41, in setup_module
    ctx2 = cuda.Context.from_numba(nb_ctx2)
  File "pyarrow/_cuda.pyx", line 84, in pyarrow._cuda.Context.from_numba
    handle=context.handle.value)
AttributeError: 'cuda.bindings.driver.CUcontext' object has no attribute 'value'
```

This probably means that our Numba interoperation code doesn't work with Numba-CUDA, only with the deprecated CUDA support in Numba itself.

### Component(s)

GPU, Python

Oops, I forgot to write up the remaining issue - it is:

```
______________________________________ test_numba_memalloc[float32-numba.cuda] _______________________________________

c = 1, dtype = dtype('float32')

    @pytest.mark.parametrize("c", range(len(context_choice_ids)),
                             ids=context_choice_ids)
    @pytest.mark.parametrize("dtype", dtypes, ids=dtypes)
    def test_numba_memalloc(c, dtype):
        ctx, nb_ctx = context_choices[c]
        dtype = np.dtype(dtype)
        # Allocate memory using numba context
        # Warning: this will not be reflected in pyarrow context manager
        # (e.g bytes_allocated does not change)
        size = 10
        mem = nb_ctx.memalloc(size * dtype.itemsize)
        darr = DeviceNDArray((size,), (dtype.itemsize,), dtype, gpu_data=mem)
        darr[:5] = 99
        darr[5:] = 88
        np.testing.assert_equal(darr.copy_to_host()[:5], 99)
        np.testing.assert_equal(darr.copy_to_host()[5:], 88)
    
        # wrap numba allocated memory with CudaBuffer
>       cbuf = cuda.CudaBuffer.from_numba(mem)

pyarrow/tests/test_cuda_numba_interop.py:178: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   if mem.device_pointer.value is None and mem.size==0:
E   AttributeError: 'cuda.bindings.driver.CUdeviceptr' object has no attribute 'value'
```

`memalloc` returns `c_void_p` objects with the ctypes bindings, and `CUdeviceptr` objects with the NVIDIA bindings. This is obviously an unpleasant inconsistency, but I've not yet thought of a straightforward fix because other libraries had adapted to this inconsistency (e.g. https://github.com/rapidsai/rmm), so just straightforwardly making it consistent will break them.

--- a/.env
+++ b/.env
@@ -67,6 +67,7 @@ LLVM=18
 MAVEN=3.8.7
 NODE=20
 NUMBA=latest
+NUMBA_CUDA=latest
 NUMPY=latest
 PANDAS=latest
 PYTHON=3.9

--- a/ci/docker/linux-apt-python-3.dockerfile
+++ b/ci/docker/linux-apt-python-3.dockerfile
@@ -32,9 +32,10 @@ RUN python3 -m venv ${ARROW_PYTHON_VENV} && \
       -r arrow/python/requirements-test.txt
 
 ARG numba
+ARG numba_cuda
 COPY ci/scripts/install_numba.sh /arrow/ci/scripts/
 RUN if [ "${numba}" != "" ]; then \
-        /arrow/ci/scripts/install_numba.sh ${numba} \
+        /arrow/ci/scripts/install_numba.sh ${numba} ${numba_cuda} \
     ; fi
 
 ENV ARROW_ACERO=ON \

--- a/ci/scripts/install_numba.sh
+++ b/ci/scripts/install_numba.sh
@@ -19,8 +19,8 @@
 
 set -e
 
-if [ "$#" -ne 1 ]; then
-  echo "Usage: $0 <numba version>"
+if [ "$#" -ne 1 ] && [ "$#" -ne 2 ]; then
+  echo "Usage: $0 <numba version> [numba-cuda version]"
   exit 1
 fi
 
@@ -40,4 +40,18 @@ elif [ "${numba}" = "latest" ]; then
   pip install numba
 else
   pip install "numba==${numba}"
+fi
+
+if [ "$#" -eq 1 ]; then
+  exit 0
+fi
+
+numba_cuda=$2
+
+if [ "${numba_cuda}" = "master" ]; then
+  pip install https://github.com/NVIDIA/numba-cuda/archive/main.tar.gz#egg=numba-cuda
+elif [ "${numba_cuda}" = "latest" ]; then
+  pip install numba-cuda
+else
+  pip install "numba-cuda==${numba_cuda}"
 fi

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -560,10 +560,11 @@ services:
 
   ubuntu-cuda-cpp:
     # Usage:
-    #   docker compose build cuda-cpp
-    #   docker compose run --rm cuda-cpp
+    #   docker compose build ubuntu-cuda-cpp
+    #   docker compose run --rm ubuntu-cuda-cpp
     # Parameters:
     #   ARCH: amd64
+    #   UBUNTU: 22.04, 24.04
     #   CUDA: <depends on your nvidia driver, should match system CUDA>
     image: ${REPO}:${ARCH}-ubuntu-${UBUNTU}-cuda-${CUDA}-cpp
     build:
@@ -946,13 +947,15 @@ services:
 
   ubuntu-cuda-python:
     # Usage:
-    #   docker compose build cuda-cpp
-    #   docker compose build cuda-python
-    #   docker compose run --rm cuda-python
+    #   docker compose build ubuntu-cuda-cpp
+    #   docker compose build ubuntu-cuda-python
+    #   docker compose run --rm ubuntu-cuda-python
     # Parameters:
     #   ARCH: amd64
     #   CUDA: <depends on your nvidia driver, should match system CUDA>
     #   UBUNTU: 22.04, 24.04
+    #   NUMBA: master, latest, <version>
+    #   NUMBA_CUDA: master, latest, <version>
     image: ${REPO}:${ARCH}-ubuntu-${UBUNTU}-cuda-${CUDA}-python-3
     build:
       context: .
@@ -962,6 +965,7 @@ services:
       args:
         base: ${REPO}:${ARCH}-ubuntu-${UBUNTU}-cuda-${CUDA}-cpp
         numba: ${NUMBA}
+        numba_cuda: ${NUMBA_CUDA}
     shm_size: *shm-size
     environment:
       <<: [*common, *ccache, *sccache]

--- a/python/pyarrow/_cuda.pyx
+++ b/python/pyarrow/_cuda.pyx
@@ -451,9 +451,9 @@ cdef class CudaBuffer(Buffer):
           Device buffer as a view of numba MemoryPointer.
         """
         ctx = Context.from_numba(mem.context)
-        if mem.device_pointer.value is None and mem.size==0:
+        if mem.device_pointer_value is None and mem.size==0:
             return ctx.new_buffer(0)
-        return ctx.foreign_buffer(mem.device_pointer.value, mem.size, base=mem)
+        return ctx.foreign_buffer(mem.device_pointer_value, mem.size, base=mem)
 
     def to_numba(self):
         """Return numba memory pointer of CudaBuffer instance.

